Always faster and always more accurate monitor rendering with Iris.

toad-dev opened the pull request:
Depends on #96. Merge that first.

- In order to fix a slight visual artifact caused by using a z-offset on monitors, I split the terminal geometry into a foreground and background VBO and made the foreground render with glPolygonOffset. This effectively makes everything in plane without z-fighting. Having two VBOs does slow things down a little bit.
- To speed things back up I added an Iris integration so we can detect when we're in the shadow pass and use simplified geometry. This can double frame rates when rendering really complex monitor geometry.

toad-dev committed:
Use polygon offset on foreground monitor geometry.

This gets rid of a visual artifact where you could sometimes see the
background of a terminal character peeking out from the edge when it
shouldn't be visible.

toad-dev committed:
Don't render complex monitor screen geometry during shadow pass.

- Add an interface to Iris's api.

- Use that interface to render empty terminals during the shadow pass.
  Empty terminals give the same depth information but are much cheaper
  to render. Not every shaderpack enables shadows for block entities,
  but for the ones that do this can drastically speed things up. I saw
  an improvement from 20 ms to 11 ms per frame with BSL shaders when
  rendering 120 static monitors filled with noise.

toad-dev commented:
We should probably make the main branch 1.18.2.

toad-dev committed:
Use a single VBO and draw with offsets.

- Record the background and foreground vertex counts while drawing the
  vbo geometry.

- Use these counts to draw background and foreground geometry from a
  single VBO with glDrawElementsBaseVertex.

- The profiler shows a ~10% improvement in time spent drawing monitors.
  For reasons unknown to me, the gain is all in the draw calls and not
  the vbo uploading. ¯\_(ツ)_/¯

toad-dev commented:
Managed to get it back down to one VBO with two draw calls. Also rewrote the commit history so it's easier to A/B test the changes.

SquidDev commented on src/main/java/dan200/computercraft/shared/integration/IrisCompat.java:
```diff
@@ -0,0 +1,72 @@
+
+    public static ApiWrapper api = new ApiWrapper() {};
+
+    public static void initialise()
```
I wonder if it'd be better to do this during the static initialiser, so we can make the field `static final`.

Honestly, I'd be more inclined to write this a bit like:

```java
public class IrisCompat
{
    public static final IrisCompat INSTANCE = FabricLoader.getInstance().isModLoaded( "iris" ) ? new IrisCompat() : new Impl();

    public boolean isRenderingShadowPass()
    {
        return false;
    }

    private static class Impl
    {
        @Override
        public boolean isRenderingShadowPass()
        {
            return false;
        }
    }
}
```

Not sure if we want to drop the other methods too - not sure what they're used for!

There's a bit of me which wonders if we could/should make this generic for other shader mods (for instance, Optifine has a `net.optifine.Shaders.isShadowPass` field), but you did mention that Optifabric is still broken, so possibly not worth it.

toad-dev replied:
That is much nicer, thank you. I think for this cut I'm going to ignore Optifine. I do want to revisit it in the future though.

SquidDev commented on src/main/java/dan200/computercraft/client/render/TileEntityMonitorRenderer.java:
```diff
@@ -235,22 +225,19 @@ private static void renderTerminal( @Nonnull MultiBufferSource bufferSource, Pos
                 // Render foreground geometry with glPolygonOffset enabled. Note we don't care about drawing these in
                 // front to back order because it's a cutout type shader that uses discard.
-                bufferSource.getBuffer( RenderTypes.MONITOR_POLYGON_OFFSET );
                 RenderTypes.MONITOR_POLYGON_OFFSET.setupRenderState();
```
I wonder if we can get rid of the `bufferSource.getBuffer( RenderType.solid() );` now (or at least in the VBO case). We don't ever touch `MultiBufferSource`. Might save a few draw calls.

I also wonder if it's better doing the polygon offset stuff ourself rather than using render types. In principle I'm not a fan, but it would reduce some work that `setupRenderState` does - worth benchmarking maybe?

toad-dev replied:
I tried to removing `bufferSource.getBuffer( RenderType.solid() );` but everything went haywire! Monitor geometry was smearing across the screen. I might be missing a call BufferUploader.reset() somewhere...

`setupRenderState` was showing up in my profiles as the 2nd heaviest method behind the `DirectVertexBuffer`'s draw routines, so I went ahead and setup `glPolygonOffset` directly. It makes a difference in the profile and IMO seems harmless to do directly.

SquidDev replied:
Aww, that's sad, but fair enough!

SquidDev commented on src/main/java/dan200/computercraft/client/render/text/DirectFixedWidthFontRenderer.java:
```diff
@@ -165,19 +165,24 @@ public static void drawTerminalBackground(
         );
-        drawTerminalBackground(
+
+        int pivot = buffer.position();
```
Instead of returning the pivot, I'd honestly just call the individual `drawTerminalForeground`/`drawTerminalBackground` methods in `TileEntityMonitorRenderer`.

toad-dev replied:
Done :+1:

SquidDev commented on src/main/java/dan200/computercraft/client/util/DirectVertexBuffer.java:
```diff
@@ -48,11 +51,85 @@ public void upload( int vertexCount, VertexFormat.Mode mode, VertexFormat format
+        indexCount = actualIndexCount;
+    }
+
+    private void drawWithShaderBaseVertex( Matrix4f matrix4f, Matrix4f matrix4f2, ShaderInstance shaderInstance, int baseVertex )
```
I think it's probably worth making `DirectVertexBuffer` its own class rather than inheriting from `VertexBuffer`. I originally subclassed in an attempt to be compatible with anybody who mixed into `drawWithShader`[^1], but seeing as we're overriding everything now it doesn't make much sense.

Will mean we can drop some access transformers too!


[^1]: Optifine. Iris doesn't appear to.

toad-dev replied:
I did this and then went a little crazy in the last commit trying to tailor `DirectVertexBuffer` to our specific use case in order to remove as much redundant setup as possible. Maybe it's too much? Let me know what you think.

SquidDev replied:
:+1: Probably worth removing the access widener lines, but otherwise looks very good!

toad-dev replied:
Removed the Access wideners :+1:

SquidDev reviewed (changes requested):
:heart: :+1:

toad-dev committed:
Don't subclass VertexBuffer.

There's no reason to subclass now that we don't need to call
VertexBuffer.drawWithShader(). This first cut is set up like the Mojang
VertexBuffer, except with our own gl*BaseVertex drawing routine.

toad-dev committed:
Reduce cpu overhead in VBO backend.

- Avoid doubling up on state setup by adding a begin() and end() method
  to DirectVertexBuffer. In between calls to these methods, we can draw
  out of the vbo multiple times.

- Avoid setting up the VAO state every frame by using a vaoInitialised
  flag. My (possibly incorrect!) understanding of VAOs is that they're
  supposed to be used as persistent hold of all state centered around
  supplying vertices. We can set up the attribute vertex pointers once,
  stop binding/unbinding the vertex buffer, and keep the element array
  binding in the VAO state. I don't know why Mojang's version redoes all
  this constantly.

- Finally, reduce setupRenderState() work by setting up glPolygonOffset
  ourselves.

These changes make an obvious difference in profiles, but are really
hard to make apparent in game, at least on my machine. I'm perhaps not
cpu bound enough. So, to push my cpu I made a world with 10,240 1x1
monitors, and saw a savings of ~7ms per frame (17fps -> 21fps). That
is so far from a real world test though, that I wonder if any of this
is valuable to users?

SquidDev commented:
> Avoid setting up the VAO state every frame by using a vaoInitialised flag.

Oooooh, nice.

> That is so far from a real world test though, that I wonder if any of this is valuable to users?

Yes and no. At this point the gains are marginal, so a lot of this is optimisation for the fun of it (well, I've had fun working on monitors - obviously cannot speak for you!). That said, if every mod can shave off a few % of their rendering time, it does all add up and result in tangible FPS improvements!

SquidDev commented:
Possibly worth doing a `git rebase --onto` as well, just so clean up some of the conflicts with #96.

toad-dev committed:
Don't force a buffer flush on VBO backend.

Only the TBO backend needs this now. This didn't work previously, but I
guess with the latest changes to DirectVertexBuffer it isn't leaking
state anymore. Yay!

toad-dev commented:
I was thinking about the title ("sometimes faster") and decided to do some testing to see where this PR stands performance-wise. The results were unexpectedly good:

| Test Case | Description | Pre-PR | Two VBOs w/ `glPolygonOffset` | PR (One VBO + less cpu overhead) | 
|-----------|-------------|--------|-----------------------------|----|
| 1         | 480 static, noise filled monitors | 42 fps | 86 fps | 92 fps |
| 2         | 120 monitor `blit 0` | 196 fps | 175 fps | ~275 fps~* 205 fps |
| 3         | 10,240 1x1 empty monitors | 17 fps | 15 fps | 24 fps |

Test 1 is designed to be a gpu bound. Test 2 is for testing the effects of VBO building and uploading but is still very sensitive to gpu throughput. Test 3 is for hammering the cpu to expose any overhead.

All these samples were taken without Iris enabled, so the shadowpass optimization would create additional *uplift* when using most shaderpacks, though overall framerates would obviously be lower.

What I didn't expect—and the reason I included the two VBO version—is that drawing with `glPolygonOffset` seems to be significantly faster for the gpu. Even using `glPolygonOffset` in the naive way doubled static framerates and offset a huge portion of the second VBO's added overhead. I don't know why this is, but I'll take it! It's possible I was doing something very wrong before with the z-offset method.

So anyway it's safe to say "sometimes faster" is no longer accurate.

---

*On further testing, this run was one of those lucky JIT outliers. Only about 1 in 20 startups give me this performance level.

toad-dev commented:
> well, I've had fun working on monitors - obviously cannot speak for you!

:heart: I've had a lot of fun too! I've never even played Minecraft with shaders, but watching the numbers go up and knowing somebody is gonna have a better experience is satisfying.

SquidDev reviewed (approved):
This is incredible work, :+1:. Wow, wasn't expecting performance to be better at all at the start of this, let alone _that_ much better.

Merith-TK commented:
For the next day or two I won't be available for this, if you feel comfortable with it, you should have merge permissions Toad, if not, squid can grant you them as the only access to GH I will have is through the app,

Carrying out part of an family member's will,

SquidDev commented:
Oh gosh. Hope you're doing okay!

Merith-TK commented:
doing okay, just *traffic* is overwhelming in the Seattle area

Merith-TK commented:
is this ready to merge???

SquidDev commented:
Yep.
